kafka-rest/src/main/java/io/confluent/kafkarest/backends/kafka/KafkaModule.java

```diff
@@ -121,4 +127,20 @@ public void dispose(Producer<byte[], byte[]> producer) {
       // Producer is disposed when the KafkaRestContext is disposed.
     }
   }
+
+  private static final class ProducerMetricsFactory implements Factory<ProducerMetrics> {
+
+    private ProducerMetrics producerMetrics;
+
+    @Override
+    public ProducerMetrics provide() {
+      producerMetrics = new ProducerMetrics(Time.SYSTEM);
+      return producerMetrics;
+    }
```

Comment:
Given that this is used for DI purposes in a mostly concurrent matter, it's an interesting question how safe is this publishing. On first glance, `ProducerMetrics` is kinda safely-initialized, as it uses `final` fields, but the objects referred by the fields are actually mutable and are being mutated after the fields are initialized. We should ponder a bit more whether there might be a problem around that, although I think just not treating this as safely-initialized, and publishing via `volatile` would be good enough.

Reply:
as per discussion on slack I've made producerMetrics volatile and guarded it with an IllegalStateException if provide() is called with a non null metrics object. I've manually tested this with multiple connections but not with multiple contexts, I believe that test would be in ce-kafka-rest as we don't do multi context in kafka-rest.